rosbag2_storage/include/rosbag2_storage/storage_factory.hpp

```diff
@@ -0,0 +1,52 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_STORAGE__STORAGE_FACTORY_HPP_
+#define ROSBAG2_STORAGE__STORAGE_FACTORY_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rosbag2_storage/storage_interface.hpp"
+#include "rosbag2_storage/visibility_control.h"
+
+namespace rosbag2_storage
+{
+
+class StorageFactoryImpl;
+
+/// Factory to create instances for various storage interfaces
+class StorageFactory
+{
+public:
+  explicit StorageFactory(const std::string & storage_identifier);
+
+  virtual ~StorageFactory();
+
+  /// Static function to create \sa StorageInterface given the storage_identifier
+  /**
+   * The function returns the StorageInterface base class for a specific identifier.
+   * \param storage_identifier unique string indicating which storage format to use.
+   */
+  ROSBAG2_STORAGE_PUBLIC
+  std::shared_ptr<StorageInterface>
+  get_storage_interface();
```

Comment:
The factory instance gets the identifier and thus `get_storage_interface()` shall only return one specific type of interface. But you're right, that might be misleading. We could maybe template the class with the interface type to have it a strongly typed factory?

Reply:
Well at least I would update the documentation, since it says that this function takes the identifier. I guess a template argument is ok. I don't have the context to say if it's required/desired or not.